### PR TITLE
types: parse string to time compatibility

### PR DESCRIPTION
**Pull request by edytagarbarz**
### What problem does this PR solve? 
#11866 

### What is changed and how it works?

- allow leading and trailing white spaces
- invalid last part of time causes an error 
```sh
mysql> select time("10:11:1212");
ERROR 1105 (HY000): expected space in input to match format
```
similarly to
```sh
mysql> select time("10:112:12");
ERROR 1105 (HY000): input does not match format
``` 
- invalid leading day value causes an error
```sh
mysql> select time("-1? 10:11:1212");
+------------------------+
| time("-1? 10:11:1212") |
+------------------------+
| 00:00:00               |
+------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------+
| Level   | Code | Message                                          |
+---------+------+--------------------------------------------------+
| Warning | 1292 | Truncated incorrect time value: '-1? 10:11:1212' |
+---------+------+--------------------------------------------------+
1 row in set (0.00 sec)
```

### Check List 

Tests 

 - Unit test
- Changed the integration test for 'bad string' in TIME_FORMAT function. Documentation doesn't specify whether NULL or 0 should be returned but if 0 is preferred here, it looks like the problem is in TimeFormat and someone else's opinion would be appreciated

### CODE REVIEW

**Comment by edytagarbarz** [comment]
@qw4990 , thanks for the reply. Sure, but imo it should also be the case for the other problems with format  e.g. invalid middle part. Currently an error is returned and there are integration tests that are based on that. What do you think?

**Comment by qw4990** [comment]
> @qw4990 , thanks for the reply. Sure, but imo it should also be the case for the other problems with format e.g. invalid middle part. Currently an error is returned and there are integration tests that are based on that. What do you think?

Seems you are right, and I create [a new issue for this problem](https://github.com/pingcap/tidb/issues/14509).

File: types/time.go [hunk 1]
```diff
@@ -1214,11 +1214,26 @@ func ParseDuration(sc *stmtctx.StatementContext, str string, fsp int8) (Duration
...
+	str = strings.TrimSpace(str)
+
+	// Time format may have day.
 	if n := strings.IndexByte(str, ' '); n >= 0 {
-		if day, err = strconv.Atoi(str[:n]); err == nil {
-			dayExists = true
+		day, err = strconv.Atoi(str[:n])
+		// Time format may have date
+		if err != nil {
+			t, err1 := ParseDatetime(sc, str)
```
Comment (qw4990) [inline types/time.go]: Could you please explain why call `ParseDatetime` here 👀?

Comment (edytagarbarz) [reply types/time.go]: Please have a look at this examples:
"1ad 10:10:10" - this one is not correct
"2019-01-12 10:10:10" - this one is correct
```sh
mysql> select time("1ad 10:13:12");
+----------------------+
| time("1ad 10:13:12") |
+----------------------+
| NULL                 |
+----------------------+
1 row in set, 1 warning (0.01 sec)

mysql> select time("2012-12-11 10:13:12");
+-----------------------------+
| time("2012-12-11 10:13:12") |
+-----------------------------+
| 10:13:12                    |
+-----------------------------+
1 row in set (0.00 sec)
```
but in both cases strconv.Atoi returns an error. Previously the error was dismissed and 1st example was considered as correct (just without a day)

File: types/time.go [hunk 2]
```diff
@@ -1261,6 +1276,8 @@ func ParseDuration(sc *stmtctx.StatementContext, str string, fsp int8) (Duration
 				_, err = fmt.Sscanf(integeralPart, "%2d", &second)
 			case 1: // 0S
 				_, err = fmt.Sscanf(integeralPart, "%1d", &second)
+			case 0:
+				return ZeroDuration, nil
```
Comment (wshwsh12) [inline types/time.go]: Why add these lines ?

File: types/time.go (hunk 1 above)
Comment (qw4990) [reply types/time.go]: got.

File: types/time.go (hunk 2 above)
Comment (qw4990) [inline types/time.go]: `time(" ")` returns null in MySQL while this line let TiDB return the zero duration.

**Pull request closed without merging**